synapse/lib/stormlib/imap.py

```diff
@@ -1,30 +1,463 @@
+    '''
+    def on_error(exc):
+        # Escaped double-quote or backslash not in quotes
+        if exc.token_history and len(exc.token_history) == 1 and (tok := exc.token_history[0]).type == 'UNQUOTED_CHAR' and tok.value == '\\':
+            mesg = f'Invalid data: {exc.token.value} cannot be escaped.'
+            raise s_exc.BadDataValu(mesg=mesg, data=text) from None
+
+        if exc.token.type == 'UNQUOTED_CHAR' and exc.expected == {'QUOTED_SPECIALS'}:
+            mesg = f'Invalid data: {exc.token.value} cannot be escaped.'
+            raise s_exc.BadDataValu(mesg=mesg, data=text) from None
+
+        # Double quote (opening a quoted string) at end of line
+        if exc.token.type == 'DBLQUOTE' and exc.column == len(text):
+            mesg = 'Quoted strings must be preceded and followed by a space.'
+            raise s_exc.BadDataValu(mesg=mesg, data=text) from None
+
+        # Unclosed quoted string
+        if exc.token.type == '$END' and exc.column == len(text) and exc.expected == {'QUOTED_CHAR', 'DBLQUOTE', 'BACKSLASH'}:
+            mesg = 'Unclosed quotes in text.'
+            raise s_exc.BadDataValu(mesg=mesg, data=text) from None
+
+        # Catch-all exception
+        raise s_exc.BadDataValu(mesg='Unable to parse IMAP response data.', data=text) from None # pragma: no cover
+
+    tree = LarkParser.parse(text, on_error=on_error)
+    newtree = AstConverter(text).transform(tree)
+    return newtree.children
+
+imap_rgx = regex.compile(
+    br'''
+    ^
+      (?P<tag>\*|\+|[0-9a-zA-Z]+)       # tag is mandatory
+      (\s(?P<uid>[0-9]+))?              # uid is optional
+      (\s(?P<response>[A-Z]{2,}))       # response is mandatory
+      (\s\[(?P<code>.*?)\])?            # code is optional
+      (\s(?P<data>.*?(?! {\d+})))?      # data is optional
+      (\s({(?P<size>\d+)}))?            # size is optional
+    $
+    ''',
+    flags=regex.VERBOSE
+)
+
+imap_rgx_cont = regex.compile(
+    br'''
+    ^
+      ((?P<data>.*?(?! {\d+})))?        # data is optional
+      (\s({(?P<size>\d+)}))?            # size is optional
+    $
+    ''',
+    flags=regex.VERBOSE
+)
+
+class IMAPBase(s_link.Link):
+    '''
+    Base class for IMAPClient and IMAPServer (in test_lib_stormlib_imap.py).
+    '''
+    async def __anit__(self, reader, writer, info=None, forceclose=False):
+        await s_link.Link.__anit__(self, reader, writer, info=info, forceclose=forceclose)
+
+        self._rxbuf = b''
+        self.state = 'LOGOUT'
+
+    def _parseLine(self, line): # pragma: no cover
+        raise NotImplementedError('Not implemented')
+
+    def pack(self, mesg): # pragma: no cover
+        raise NotImplementedError('Not implemented')
+
+    def feed(self, byts):
+        ret = []
+
+        # Append new bytes to existing bytes
+        self._rxbuf += byts
+
+        # Iterate through buffer and parse out complete messages
+        while (offs := self._rxbuf.find(CRLF)) != -1:
```

Comment:
We have two while loops here - can we end up in a situation where they never break / return & result in a task that cannot be cancelled or interrupted?

Reply:
The only way that could happen is if the server was continuously streaming messages (lines). That situation should never happen though. I could probably make it return after a fixed number of messages (10 or something?).

Reply:
Limited it to 32 messages.